uniswap/uniswap4.py

```diff
@@ -0,0 +1,788 @@
+
+    def _build_and_send_tx(
+        self, function: ContractFunction, tx_params: Optional[TxParams] = None
+    ) -> HexBytes:
+        """Build and send a transaction."""
+        if not tx_params:
+            tx_params = self._get_tx_params()
+        transaction = function.build_transaction(tx_params)
+        # Uniswap3 uses 20% margin for transactions
+        transaction["gas"] = Wei(int(self.w3.eth.estimate_gas(transaction) * 1.2))
+        signed_txn = self.w3.eth.account.sign_transaction(
+            transaction, private_key=self.private_key
+        )
+        # TODO: This needs to get more complicated if we want to support replacing a transaction
+        # FIXME: This does not play nice if transactions are sent from other places using the same wallet.
+        try:
+            return self.w3.eth.send_raw_transaction(signed_txn.rawTransaction)
+        finally:
+            logger.debug(f"nonce: {tx_params['nonce']}")
+            self.last_nonce = Nonce(tx_params["nonce"] + 1)
+
+    def _get_tx_params(self, value: Wei = Wei(0), gas: Optional[Wei] = None, max_fee: Optional[Wei] = None, priority_fee: Optional[Wei] = None) -> TxParams:
+        """Get generic transaction parameters."""
+        params: TxParams = {
+            "from": _addr_to_str(self.address),
+            "value": value,
+            "nonce": max(
+                self.last_nonce, self.w3.eth.get_transaction_count(self.address)
+            ),
+        }
+
+        if gas:
+            params["gas"] = gas
+        if max_fee:
+            params["maxFeePerGas"] = max_fee
+        if priority_fee:
+            params["maxPriorityFeePerGas"] = priority_fee
+
+        return params
+
+    # ------ Helpers ------------------------------------------------------------
+
+    def get_token(self, address: AddressLike, abi_name: str = "erc20") -> ERC20Token:
+        """
+        Retrieves metadata from the ERC20 contract of a given token, like its name, symbol, and decimals.
+        """
+        # FIXME: This function should always return the same output for the same input
+        #        and would therefore benefit from caching
+        if address == ETH_ADDRESS:
+            return ERC20Token("ETH", address, "Ether", 18)
+        token_contract = _load_contract(self.w3, abi_name, address=address)
+        try:
+            _name = token_contract.functions.name().call()
+            _symbol = token_contract.functions.symbol().call()
+            decimals = token_contract.functions.decimals().call()
+        except Exception as e:
+            logger.warning(
+                f"Exception occurred while trying to get token {_addr_to_str(address)}: {e}"
+            )
+            raise InvalidToken(address)
+        try:
+            name = _name.decode()
+        except:
+            name = _name
+        try:
+            symbol = _symbol.decode()
+        except:
+            symbol = _symbol
+        return ERC20Token(symbol, address, name, decimals)
```

Comment:
Can we reuse this from the other class? `Uniswap.get_token(self, address, abi_name)`?

Many such cases, not sure what to do about them all.

Reply:
Specifically this function might be dropped, it isn't in use now. It is worth considering removing everything, that isn't related to the low level itself.

Reply:
Lol, I was going to do that exactly but forgot to remove this method.